Fix encoding headers in python files

xylar opened the pull request:
This merge moves the encoding headers to the top of python files.  It also updates the code for prepending license headers to make sure encoding headers go before the license header.

A few files have had license headers added to them.

xylar commented:
This fix is expected to address #318.

xylar commented:
## Testing

I ran the `license-prepend-missing.sh` script several times to make sure encoding headers (and combinations of encoding headers and shebangs) ended up above the license.

In the process, I discovered the need for the `-r` flag in `xargs` calls.

xylar commented on licenses/_headers/license-prepend-missing.sh:
```diff
@@ -16,10 +16,13 @@
 source license-setup.sh
 
+GET=( -iname "*.py" -or -iname "*.sh" -or -iname "*.bash" -or -iname "*.csh" \
+     -or -iname "*.pbs" -or -iname "*.html" -or -iname "*.css" -or -iname "*.js" )
```
@jhkennedy, this was added because we have files in our repo that don't match any of these and they were getting listed as files where a license would be added but that didn't actually happen.

jhkennedy replied:
Ah, right, b/c the find all files is more inclusive than the sum of the prepend commands. This is a good addition, but it mean `license-find-missing.sh` will also return a more inclusive list than using this script. So, it'd probably be a good idea to either: 

1. include these switches in `license-find-missing.sh` -- inserting `\( "${GET[@]}" \) ` into the find command, and including `GET=` at the top or in `license-setup.sh`

2. Keep the more inclusive find command but also print a list of files that have no rule defined to prepend the license header -- At the top of `license-prepend-missing.sh` say `ATTEMPTING TO PREPEND A LICENSE...` and then at the bottom list the files that no action was performed [possibly by just running `license-find-missing.sh` again], with a warning message to either add those files to the ignore lists or define a rule for them.

xylar commented on licenses/_headers/license-prepend-missing.sh:
```diff
@@ -29,15 +32,60 @@ find $SOURCE_DIR -type f "${ALWAYS_IGNORE[@]}" \
 find $SOURCE_DIR -type f \( "${GET[@]}" \) "${ALWAYS_IGNORE[@]}" "${PYTHON_IGNORE[@]}" \
-    | xargs grep -L "#!" \
-    | xargs grep -L "$CURRENT" \
+    | xargs -r grep -L "#!" \
```
It is safest to have a `-r` in each `xargs` call.  If the file list passed to `xargs` is empty, `xargs` returns `(standard input)`, which then causes error in the subsequent code.  With `-r`, `xargs` doesn't run and the result remains an empty string.

milenaveneziani commented:
@xylar: a test on LANL ic is working now (before it used to stop immediately with the enconding error, but now things are running smoothly).

xylar commented:
Thanks @milenaveneziani.  I'll wait and see if @jhkennedy has some comments later this week before I merge.

xylar commented:
Yes, please make all those changes.

jhkennedy commented:
@xylar I've pushed two commits which address those issues, and tested the prepend script on the development branch of MPAS-Dev/MPAS-Analysis. Give them a whirl, and let me know if you see any issues.

xylar commented:
@jhkennedy, these changes look great to me.  Based on a little testing I did, they seem to work.

I rebased onto the current develop, just to be on the safe side (wasn't sure if any licenses would be affected) and I added a license to run_mpas_analysis manually by copying it from elsewhere.

Seems ready to merge as soon as CI finishes successfully.

vanroekel commented:
confirmed, fixed by #319
